File: spare-parts-shop/src/main/java/com/spareparts/model/Warranty.java
package com.spareparts.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import java.time.LocalDate;
import java.time.LocalDateTime;

@Entity
@Table(name = "warranties")
@Data
@NoArgsConstructor
@AllArgsConstructor
@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
public class Warranty implements BelongsToBusiness {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "business_id", nullable = false)
    @JsonIgnore
    private Business business;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "branch_id")
    private Branch branch;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "invoice_id", nullable = false)
    private Bill bill;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "product_id", nullable = false)
    private Product product;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "customer_id", nullable = false)
    private Customer customer;

    @Column(name = "serial_number")
    private String serialNumber;

    @Column(name = "model_number")
    private String modelNumber;

    @Column(name = "warranty_type", nullable = false)
    private String warrantyType = "NO_WARRANTY"; // NO_WARRANTY, MANUFACTURER, SELLER, EXTENDED, CUSTOM

    @Column(name = "warranty_start_date", nullable = false)
    private LocalDate warrantyStartDate;

    @Column(name = "warranty_end_date", nullable = false)
    private LocalDate warrantyEndDate;

    @Column(name = "warranty_period_months")
    private Integer warrantyPeriodMonths;

    @Column(name = "warranty_notes")
    private String warrantyNotes;

    @Column(name = "warranty_terms")
    private String warrantyTerms;

    @Column(name = "created_at")
    private LocalDateTime createdAt = LocalDateTime.now();
}

File: spare_parts_shop_flutter/lib/models/warranty.dart

import 'bill.dart';
import 'product.dart';
import 'customer.dart';

class Warranty {
  final int id;
  final Bill bill;
  final Product product;
  final Customer customer;
  final String? serialNumber;
  final String? modelNumber;
  final String warrantyType;
  final String warrantyStartDate;
  final String warrantyEndDate;
  final int? warrantyPeriodMonths;
  final String? warrantyNotes;
  final String? warrantyTerms;
  final String createdAt;

  Warranty({
    required this.id,
    required this.bill,
    required this.product,
    required this.customer,
    this.serialNumber,
    this.modelNumber,
    required this.warrantyType,
    required this.warrantyStartDate,
    required this.warrantyEndDate,
    this.warrantyPeriodMonths,
    this.warrantyNotes,
    this.warrantyTerms,
    required this.createdAt,
  });

  factory Warranty.fromJson(Map<String, dynamic> json) {
    return Warranty(
      id: json['id'],
      bill: Bill.fromJson(json['bill']),
      product: Product.fromJson(json['product']),
      customer: Customer.fromJson(json['customer']),
      serialNumber: json['serialNumber'],
      modelNumber: json['modelNumber'],
      warrantyType: json['warrantyType'],
      warrantyStartDate: json['warrantyStartDate'],
      warrantyEndDate: json['warrantyEndDate'],
      warrantyPeriodMonths: json['warrantyPeriodMonths'],
      warrantyNotes: json['warrantyNotes'],
      warrantyTerms: json['warrantyTerms'],
      createdAt: json['createdAt'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'bill': bill.toJson(),
      'product': product.toJson(),
      'customer': customer.toJson(),
      'serialNumber': serialNumber,
      'modelNumber': modelNumber,
      'warrantyType': warrantyType,
      'warrantyStartDate': warrantyStartDate,
      'warrantyEndDate': warrantyEndDate,
      'warrantyPeriodMonths': warrantyPeriodMonths,
      'warrantyNotes': warrantyNotes,
      'warrantyTerms': warrantyTerms,
      'createdAt': createdAt,
    };
  }
}
